File: GuideObjC/GuideObjC/Managers/SimpleFileManager.m
//
//  SimpleFileManager.m
//  GuideObjC
//
//  Created by LuzanovRoman on 14.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "SimpleFileManager.h"

@implementation SimpleFileManager

+ (SimpleFileManager *)sharedInstance {
    static SimpleFileManager * sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[SimpleFileManager alloc] init];
    });
    return sharedInstance;
}

- (id)init {
    self = [super init];
    
    if (self) {
        [[NSFileManager defaultManager] createDirectoryAtPath:[self cachesPath] withIntermediateDirectories:false attributes:nil error:nil];
    }
    return self;
}

- (NSString *)cachesPath {
    return [[NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES) objectAtIndex:0] stringByAppendingPathComponent:@"EGF2Images"];
}

- (void)imageWithFile:(EGFFile *)file completion:(void (^) (UIImage *image, BOOL fromCache))completion {
    if (!file || !file.id || !file.url) {
        completion(nil, false);
        return;
    }
    NSURL * url = [[NSURL alloc] initWithString:file.url];
    NSURL * localUrl = [NSURL fileURLWithPath:[[self cachesPath] stringByAppendingFormat:@"/%@",file.id]];

    if ([localUrl checkResourceIsReachableAndReturnError:nil]) {
        dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
            NSData * data = [NSData dataWithContentsOfURL:localUrl];
            UIImage * image = [UIImage imageWithData:data];
            
            dispatch_async(dispatch_get_main_queue(), ^(void){
                completion(image, true);
            });
        });
    }
    else {
        NSURLSessionConfiguration * configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
        configuration.URLCache = nil;
        configuration.HTTPCookieStorage = nil;
        NSURLSession * session = [NSURLSession sessionWithConfiguration:configuration];
        // TODO use [NSURLSession sharedSession] instead
        [[session dataTaskWithURL:url completionHandler:^(NSData * data, NSURLResponse * response, NSError * error) {
            dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
                
                UIImage * image = [UIImage imageWithData:data];
                [data writeToURL:localUrl atomically:false];
                
                dispatch_async(dispatch_get_main_queue(), ^(void){
                    completion(image, false);
                });
            });
        }] resume];
    }
}

- (void)deleteAllFiles {
    [[NSFileManager defaultManager] removeItemAtPath:[self cachesPath] error:nil];
    [[NSFileManager defaultManager] createDirectoryAtPath:[self cachesPath] withIntermediateDirectories:false attributes:nil error:nil];
}
@end

File: GuideSwift/GuideSwift/Managers/SimpleFileManager.swift
//
//  SimpleFileManager.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 13.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit

class SimpleFileManager {
    static let shared = SimpleFileManager()
    lazy var cachesURL : URL = {
        return FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).last!.appendingPathComponent("EGF2Images")
    }()
    
    init() {
        try? FileManager.default.createDirectory(at: cachesURL, withIntermediateDirectories: false, attributes: nil)
    }
    
    fileprivate func doesFileExist(withUrl url: URL) -> Bool {
        do {
            return try url.checkResourceIsReachable()
        }
        catch {
            return false
        }
    }
    
    fileprivate func fileData(withUrl url: URL) -> Data? {
        do {
            return try Data(contentsOf: url)
        }
        catch {
            return nil
        }
    }
    
    func image(withFile file: EGFFile, completion: @escaping (_ image: UIImage?, _ fromCache: Bool) -> Void) {
        guard let urlString = file.url, let fileId = file.id, let url = URL(string: urlString) else {
            completion(nil, false)
            return
        }
        let localURL = cachesURL.appendingPathComponent(fileId)
        
        func handle(data: Data?, isCached: Bool) {
            if let imageData = data, let image = UIImage(data: imageData) {
                DispatchQueue.main.async { completion(image, isCached) }
                
                if !isCached {
                    try? imageData.write(to: localURL)
                }
            }
            else {
                DispatchQueue.main.async { completion(nil, isCached) }
            }
        }
        
        if doesFileExist(withUrl: localURL) {
            DispatchQueue.global().async {
                handle(data: self.fileData(withUrl: localURL), isCached: true)
            }
        }
        else {
            let configuration = URLSessionConfiguration.default
            configuration.urlCache = nil
            configuration.httpCookieStorage = nil
            let session = URLSession(configuration: configuration)
            // TODO use URLSession.shared
            session.dataTask(with: url, completionHandler: {(data, response, error) in
                DispatchQueue.global().async {
                    handle(data: data, isCached: false)
                }
            }).resume()
        }
    }
    
    func deleteAllFiles() {
        try? FileManager.default.removeItem(at: cachesURL)
        try? FileManager.default.createDirectory(at: cachesURL, withIntermediateDirectories: false, attributes: nil)
    }
}
